Add jest to syncpack,

tomalec opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

- [Add jest to syncpack,](https://github.com/woocommerce/woocommerce/commit/bbbf15c76c3663e66795e6ca113b5099b3ffffe2) 
   as suggested in https://github.com/woocommerce/woocommerce/pull/34322#issuecomment-1226198520
- Unify the jest@^27.5.1 across the stack.
   to make `syncpack list-mismatches` happy

### How to test the changes in this Pull Request:

1. Run all automated tests
2. Run `syncpack list-mismatches`
3. change a dependency in one project, like `pnpm i --save jest-each@^26 --filter=@woocommerce/e2e-environment`
4. Run `syncpack list-mismatches` again

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

tomalec committed:
Update `jest-puppeteer@5.0.4` (from `4.4.0`)

to be albe to run with jest 28.

tomalec committed:
Update `jest` and `ts-jest@^27` (from `25`),

to match `e2e-environment` needs.

tomalec committed:
Update api-core-tests' `jest@^27` (from `25`),

to unify the used version across the stack.

tomalec committed:
Update e2e-env's `jest@^27` & co.  (from `25`),

to unify the used version across the stack.

tomalec committed:
Update e2e-env's `@jest/test-sequencer@^27`  (from `25`),

to unify the used version across the stack.

tomalec committed:
Update admin-e2e-tests' `@jest/globals@^27`  (from `26`),

to unify the used version across the stack.

tomalec committed:
Update woocommerce's `jest@^27.5.1`  (from `25`),

to unify the used version across the stack.

tomalec committed:
Replace `mocked` from `ts-jest` to the one from `jest`,

Update `@types/jest@27.4.1` from `27.0.2`.
Try to fix broken tsc build.

tomalec committed:
Merge branch 'trunk' into update/unify-jest-27

tomalec committed:
Update `e2e-core-tests`' `@jest/globals@^27.5.1`

from `^26.4.2`.

tomalec committed:
Add jest to syncpack,

as suggested in https://github.com/woocommerce/woocommerce/pull/34322#issuecomment-1226198520

tomalec committed:
Unify the `jest@^27.5.1` across the stack.

tomalec committed:
Remove changelog entry for bumpin jest in `api-core-tests`

It was already covered in `trunk`.

Addresses https://github.com/woocommerce/woocommerce/pull/34322#discussion_r956019567

tomalec committed:
Merge branch 'trunk' into update/unify-jest-27

MrJnrman committed:
Merged trunk into update/unify-jest-27

MrJnrman committed:
Updated package-lock.yaml

MrJnrman committed:
Updated timeout

tomalec committed:
Merge branch 'trunk' into update/unify-jest-27

samueljseay commented on packages/js/admin-e2e-tests/changelog/update-jest:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: update
```
The type of this changelog should be `dev`

samueljseay commented on packages/js/api/package.json:
```diff
@@ -45,14 +45,14 @@
-		"jest": "^25",
-		"ts-jest": "^25",
+		"jest": "^27.5.1",
+		"ts-jest": "^27",
```
Any reason we're not syncing ts-jest as well? with growing TS adoption it seems like we might as well get in front of that and add it to syncpack now?

tomalec replied:
I didn't add it to limit the scope of this PR, but if you say so, I'd add it as well.

samueljseay commented on packages/js/e2e-environment/changelog/update-jest-puppeteer:
```diff
@@ -1,4 +1,4 @@
 Type: update
 
-Updated `jest-puppeteer@^5.0.4`.
+Updated `jest`, `jest-circus`, `jest-each`, `@jest/test-sequencer` to `27.5.1` from `25`, and `jest-puppeteer@^5.0.4`.
```
Whats the reason for modifying an existing changelog entry? Was that on purpose?

tomalec replied:
Initially, the PR's base was `update/unify-jest-27` of https://github.com/woocommerce/woocommerce/pull/34322#issuecomment-1226198520.

I modified the existing entry in case those two PRs would go together. I can create a separate entry if that's preferred :)

samueljseay commented on packages/js/internal-js-tests/babel.config.js:
```diff
@@ -1,7 +1,3 @@
-		'@babel/preset-typescript',
-		'@wordpress/babel-preset-default',
-	],
+	presets: [ '@babel/preset-typescript', '@wordpress/babel-preset-default' ],
```
nitpick, but obviously these files have not been prettier formatted in the past, it'd be more ideal not to wholesale include formatting changes in a PR since there isn't actually a relevant change.

tomalec replied:
Interestingly, this change was introduced by commit merging the trunk to this branch (aa80b679aeec1939edd62d6036e083711249ac6e). I'll merge it again and see if it fades away.

tomalec replied:
It's no longer there after https://github.com/woocommerce/woocommerce/pull/34468/commits/a5534717418ec312c419b87955bb6c1609ad25d3

samueljseay commented on plugins/woocommerce/changelog/update-unify-jest-27:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: dev
+
+Update `jest@^26.5.1` from `^25`.
```
This description doesn't seem to match the changelog name is that intentional? Shouldn't it mention jest 27 somewhere in the description?

samueljseay reviewed (commented):
A few questions and nitpicks from me. Otherwise this is looking good.

tomalec committed:
Add separate changelog entry for `e2e-environment`,

Addresses https://github.com/woocommerce/woocommerce/pull/34468/files#r964205325

tomalec committed:
Fix a typo in the `wocommerce` changelog entry

addresses https://github.com/woocommerce/woocommerce/pull/34468/files#r964206379

tomalec committed:
Merge branch 'trunk' into add/jest-to-syncpack

tomalec committed:
Add ts-jest to syncpack

Addresses https://github.com/woocommerce/woocommerce/pull/34468#discussion_r964203794

tomalec committed:
Update `ts-jest@^27.1.3` in `api`,

Addresses https://github.com/woocommerce/woocommerce/pull/34468#discussion_r964203794

tomalec committed:
Update changelogs after merges from trunk

Addresses https://github.com/woocommerce/woocommerce/pull/34468#discussion_r964206379

tomalec commented:
@samueljseay Thank you for the review :) I addressed all your comments, would you mind another look?

tomalec committed:
Merge branch 'trunk' into add/jest-to-syncpack

tomalec committed:
Remove reduntant changelog file after merge

tomalec commented:
@samueljseay I merged the trunk again and resolved conflicts. Would you mind taking a look here?

samueljseay commented:
Yep looks like I missed the review request so long ago. apologies

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into add/jest-to-syncpack

samueljseay committed:
Fix rememo to 4.0.0 because the type def was not shipped in 4.0.1

samueljseay commented:
So unfortunately this is causing some minor version bumps that cause some cascading type issues that is still on our list to fix. I think we might have to put this on hold till I can fix these. The changes all look good, but as i try fix the type issues it becomes whack-a-mole

samueljseay committed:
Fix some minor typing issues

samueljseay commented:
Unfortunately this is quite stale and there are a lot of conflicts here so I think for now we need to close this and have another go at addressing this later.

If you are still interested or you feel like me closing this is a mistake, please open another PR based on a recent version of the trunk branch. Thanks!
